src/wh_server_nvm.c

```diff
@@ -0,0 +1,311 @@
+
+        /* Convert the response struct */
+        wh_MessageNvm_TranslateGetAvailableResponse(magic,
+                &resp, (whMessageNvm_GetAvailableResponse*)resp_packet);
+        *out_resp_size = sizeof(resp);
+    }; break;
+
+    case WH_MESSAGE_NVM_ACTION_GETMETADATA:
+    {
+        whMessageNvm_GetMetadataRequest req = {0};
+        whMessageNvm_GetMetadataResponse resp = {0};
+        whNvmMetadata meta = {0};
+
+        /* Convert request struct */
+        wh_MessageNvm_TranslateGetMetadataRequest(magic,
+                (whMessageNvm_GetMetadataRequest*)req_packet, &req);
+
+        /* Process the getmetadata action */
+        resp.rc = wh_Nvm_GetMetadata(server->nvm, req.id, &meta);
+
+        if (resp.rc == 0) {
+            resp.id = meta.id;
+            resp.access = meta.access;
+            resp.flags = meta.flags;
+            resp.len = meta.len;
+            memcpy(resp.label, meta.label, sizeof(resp.label));
+        }
+        /* Convert the response struct */
+        wh_MessageNvm_TranslateGetMetadataResponse(magic,
+                &resp, (whMessageNvm_GetMetadataResponse*)resp_packet);
+        *out_resp_size = sizeof(resp);
+    }; break;
+
+    case WH_MESSAGE_NVM_ACTION_ADDOBJECT:
+    {
+        whMessageNvm_AddObjectRequest req = {0};
+        uint16_t hdr_len = sizeof(req);
+        whNvmMetadata meta = {0};
+        const uint8_t* data = (const uint8_t*)req_packet + hdr_len;
+        whMessageNvm_SimpleResponse resp = {0};
+
+        /* Convert request struct */
+        wh_MessageNvm_TranslateAddObjectRequest(magic,
+                (whMessageNvm_AddObjectRequest*)req_packet, &req);
+
+        if(req_size == (hdr_len + req.len)) {
+            /* Process the AddObject action */
+            meta.id = req.id;
+            meta.access = req.access;
+            meta.flags = req.flags;
+            meta.len = req.len;
+            resp.rc = wh_Nvm_AddObject(server->nvm, &meta, req.len, data);
+        } else {
+            /* Problem in the request or transport. */
+            resp.rc = WH_ERROR_ABORTED;
+        }
+
+        /* Convert the response struct */
+        wh_MessageNvm_TranslateSimpleResponse(magic,
+                &resp, (whMessageNvm_SimpleResponse*)resp_packet);
+        *out_resp_size = sizeof(resp);
+   }; break;
+
+    case WH_MESSAGE_NVM_ACTION_DESTROYOBJECTS:
+    {
+        whMessageNvm_DestroyObjectsRequest req = {0};
+        whMessageNvm_SimpleResponse resp = {0};
+
+        /* Convert request struct */
+        wh_MessageNvm_TranslateDestroyObjectsRequest(magic,
+                (whMessageNvm_DestroyObjectsRequest*)req_packet, &req);
+
+        if (req.list_count <= WH_MESSAGE_NVM_MAX_DESTROY_OBJECTS_COUNT) {
+            /* Process the DestroyObjects action */
+            resp.rc = wh_Nvm_DestroyObjects(server->nvm,
+                    req.list_count, req.list);
+        } else {
+            /* Problem in transport or request */
+            resp.rc = WH_ERROR_ABORTED;
```

Comment:
Just a thought: should this be a WH_ERROR_BADARGS? Or is that just meant for function args?

Reply:
Should be aborted.  BADARGS means that the function was called with something wrong and fixing the parameters will correct it.  In this case, we end up seeing a malformed request message so we must abort.  It is likely that the caller has a bad argument passed into the request function, but this should have been caught locally and returned then.